### ISSUE TITLE
pnpm affects dependency by running unrelated `install` script

### ISSUE THREAD

**Issue by sumboid** [pnpm/pnpm#3531]
### pnpm version:
6.7.4

### Code to reproduce the issue:
It should be enough to clone https://github.com/sumboid/pnpm-broken-deps and perform `pnpm install`
In case `ws` install script gets failed you see @babel/types files are also affected: you can take a look at `./node_modules/.pnpm/@babel+types@7.14.5/node_modules/@babel/types/lib/ast-types/generated/index.js` and find node-gyp error log.

### Expected behavior:
Only `builderror.log` is affected

### Actual behavior:
Following files also contain node-gyp error output:
`./node_modules/.pnpm/@babel+types@7.14.5/node_modules/@babel/types/lib/converters/Scope.js`
`./node_modules/.pnpm/@babel+types@7.14.5/node_modules/@babel/types/lib/ast-types/generated/index.js`

### Additional information:
Environment:
* GNU/Linux
* Node 14.17.0

As far as I could investigate, pnpm makes these files to refer common inode (probably because all of them are empty at this point) and then runs `install` script of `ws` package:
```
→ ls -li ./node_modules/.pnpm/ws@0.4.32/node_modules/ws/builderror.log
134464036 -rw-r--r-- 4 user user 91387 Jun 12 12:43 ./node_modules/.pnpm/ws@0.4.32/node_modules/ws/builderror.log

→ find . -inum 134464036 
./node_modules/.pnpm/@babel+types@7.14.5/node_modules/@babel/types/lib/converters/Scope.js
./node_modules/.pnpm/@babel+types@7.14.5/node_modules/@babel/types/lib/ast-types/generated/index.js
./node_modules/.pnpm/ws@0.4.32/node_modules/ws/builderror.log

→ head ./node_modules/.pnpm/ws@0.4.32/node_modules/ws/builderror.log
gyp info it worked if it ends with ok
gyp info using node-gyp@8.1.0
gyp info using node@14.17.0 | linux | x64
gyp info find Python using Python version 3.9.5 found at "/usr/sbin/python3"
gyp info spawn /usr/sbin/python3
gyp info spawn args [
gyp info spawn args   '/home/user/.nvm/versions/node/v14.17.0/lib/node_modules/pnpm/dist/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',

→ head ./node_modules/.pnpm/@babel+types@7.14.5/node_modules/@babel/types/lib/ast-types/generated/index.js
gyp info it worked if it ends with ok
gyp info using node-gyp@8.1.0
gyp info using node@14.17.0 | linux | x64
gyp info find Python using Python version 3.9.5 found at "/usr/sbin/python3"
gyp info spawn /usr/sbin/python3
gyp info spawn args [
gyp info spawn args   '/home/user/.nvm/versions/node/v14.17.0/lib/node_modules/pnpm/dist/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
```

ws `install` script is located here: https://github.com/websockets/ws/blob/0.4.32/package.json#L24
Also if you install it ignoring scripts you may find empty `builderror.log` in the root

**Comment by sumboid** [comment]
@zkochan any ideas it is design issue or just a bug? At this point I'm not sure why pnpm links anything before running lifecycle scripts.

**Comment by zkochan** [comment]
How would pnpm run the lifecycle scripts before creating the hard links? Before the hard links are written, the package does not exist, there are just files saved at "random" locations inside the content-addressable store.

**Comment by sumboid** [comment]
@zkochan yeah I see but anyway can we consider linking "same" files to single location is expected behavior and there is nothing to fix? In general case it doesn't seem to work, it won't work even pnpm would link them within package and not across everything it manages.

**Comment by 90dy** [comment]
As temporary fix I added this to package.json:
```json
	"pnpm": {
		"overrides": {
			"ws@^0": "^1"
		}
	}
```
Don't really know if there's incompatibility between these versions.

**Comment by luckydrq** [comment]
+1

**Comment by wukd** [comment]
> How would pnpm run the lifecycle scripts before creating the hard links? Before the hard links are written, the package does not exist, there are just files saved at "random" locations inside the content-addressable store.

"install": "(node-gyp rebuild 2> builderror.log) || (exit 0)"

**Comment by wukd** [comment]
> How would pnpm run the lifecycle scripts before creating the hard links? Before the hard links are written, the package does not exist, there are just files saved at "random" locations inside the content-addressable store.

"install": "(node-gyp rebuild 2> builderror.log) || (exit 0)"
Why does this command can affect empty files in the store?

**Comment by zkochan** [comment]
This no longer reproduces on current pnpm, so closing.

Verified on pnpm 12.5.1 against the repro repo sumboid/pnpm-broken-deps: @babel/types files are intact after the install.

If you still hit this on the latest pnpm, please comment with a reproduction and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).
